### ISSUE TITLE
Node disconnects and does not reconnect, despite available peers, until manual restart

### ISSUE THREAD

**Issue by cwgoes** [tendermint/tendermint#1575]
Gaia node on `gaia-5000` disconnects (no underlying loss of network connectivity) and does not reconnect to network, despite several available peers. Restarting `gaiad` manually fixed the problem (the node reconnected to the network).

Version: 0.16.0-dev-7ef5e90

Relevant logs / `net_info` / `dump_consensus_state`: https://gist.github.com/cwgoes/6f5e1e29693bf3bdf28f3b2d760992b3

(Logs above are from when the node appeared to "disconnect", I have more logs if you need them)

**Comment by ebuchman** [comment]
Any chance you have the debug logs around this time ?

**Comment by ebuchman** [comment]
I think this might be the same bug as https://github.com/tendermint/tendermint/issues/1562

In both cases the dump_consensus_state shows we're in CommitStep (step 8).

In both cases it seems we're still connected and sending data (the rates in net_info are non-zero), but not getting any consensus information.

Is there anything special about this node ? What are the specs and where in the world is it ?

**Comment by cwgoes** [comment]
For logs, see https://gist.github.com/cwgoes/6f5e1e29693bf3bdf28f3b2d760992b3#file-gaiad-logs

Node specs: https://gist.github.com/cwgoes/ad26d4d103904878131888fc800e82ab

Cheap VPS in Canada, Ubuntu 16.04 LTS, nothing special to my knowledge.

**Comment by ebuchman** [comment]
I think this is what's happening:

It's a commit in round 3 for a block that was proposed in an earlier round and some validators were already locked on.

You're the proposer for round 3, but other validators dont hear from you and end up committing the thing they're already locked on anyways (its possible to commit a block in round R that was proposed in an earlier round, even if the proposer for R is offline). 

You see your own proposal for round 3, and you see the precommit for a different block in round 3, but the code doesn't set you up to receive the block that was committed, so your consensus hangs forever

These logs are especially telling:

```
I[05-15|03:19:19.918] Added to precommit                           module=consensus vote="Vote{28:EF0B7E6CB5E9 162664/03/2(Precommit) 9DC0B9395F81 /DAFE9C997623.../ @ 2018-05-15T03:19:12.777Z}" precommits="VoteSet{H:162664 R:3 T:2 +2/3:<nil> BA{31:_x___x__x_x___x_________x_xxxx_} map[cc9ee88fdf31e5af72d23d633a83c740c193dad1:9DC0B9395F81CB0367517E57F7B6205535091B21:1:D27A2DAA5E81]}"
I[05-15|03:19:20.186] Added to precommit                           module=consensus vote="Vote{2:22F03A4B0F6B 162664/03/2(Precommit) 9DC0B9395F81 /83F862D1E97A.../ @ 2018-05-15T03:19:12.654Z}" precommits="VoteSet{H:162664 R:3 T:2 +2/3:<nil> BA{31:_xx__x__x_x___x_________x_xxxx_} map[cc9ee88fdf31e5af72d23d633a83c740c193dad1:9DC0B9395F81CB0367517E57F7B6205535091B21:1:D27A2DAA5E81]}"
I[05-15|03:19:20.241] Added to precommit                           module=consensus vote="Vote{17:A1BDEA32E5B9 162664/03/2(Precommit) 9DC0B9395F81 /B72A5BA5A1AB.../ @ 2018-05-15T03:19:12.648Z}" precommits="VoteSet{H:162664 R:3 T:2 +2/3:9DC0B9395F81CB0367517E57F7B6205535091B21:1:D27A2DAA5E81 BA{31:_xx__x__x_x___x__x______x_xxxx_} map[cc9ee88fdf31e5af72d23d633a83c740c193dad1:9DC0B9395F81CB0367517E57F7B6205535091B21:1:D27A2DAA5E81]}"
I[05-15|03:19:20.241] enterNewRound(162664/3). Current: 162664/1/RoundStepPrecommit module=consensus 
I[05-15|03:19:20.241] enterPropose(162664/3). Current: 162664/3/RoundStepNewRound module=consensus 
I[05-15|03:19:20.241] enterPropose: Our turn to propose            module=consensus proposer=B2BF2E52EF56789AD46C326930FCE34FBE48F340 privValidator="PrivValidator{B2BF2E52EF56789AD46C326930FCE34FBE48F340 LH:162664, LR:1, LS:3}"
I[05-15|03:19:20.245] Signed proposal                              module=consensus height=162664 round=3 proposal="Proposal{162664/3 1:7F19A30606D3 (1,:0:000000000000) /01AB300ABE5F.../ @ 2018-05-15T03:19:20.242Z}"
I[05-15|03:19:20.252] enterPrecommit(162664/3). Current: 162664/3/RoundStepPropose module=consensus 
I[05-15|03:19:20.253] enterPrecommit: No +2/3 prevotes during enterPrecommit. Precommitting nil. module=consensus 
I[05-15|03:19:20.257] Signed and pushed vote                       module=consensus height=162664 round=3 vote="Vote{19:B2BF2E52EF56 162664/03/2(Precommit) 000000000000 /C49BAE5A79E8.../ @ 2018-05-15T03:19:20.253Z}" err=null
I[05-15|03:19:20.270] enterCommit(162664/3). Current: 162664/3/RoundStepPrecommit module=consensus 
I[05-15|03:19:20.274] Attempt to finalize failed. We don't have the commit block. module=consensus height=162664 proposal-block= commit-block=9DC0B9395F81CB0367517E57F7B6205535091B21
I[05-15|03:19:20.288] Received complete proposal block             module=consensus height=162664 hash=3C0555F961CEE5E150B145A5EADA10E6BD41329D
I[05-15|03:19:20.289] Attempt to finalize failed. We don't have the commit block. module=consensus height=162664 proposal-block=3C0555F961CEE5E150B145A5EADA10E6BD41329D commit-block=9DC0B9395F81CB0367517E57F7B6205535091B21
```

**Comment by ebuchman** [comment]
@jaekwon @milosevic check it out ^^ 

I don't think it should be too hard to fix, but looks like we don't handle the case where have a different proposal block than the one that was committed.

This could happen if we're the proposer in a round that commits something that was proposed earlier, or if there's a byzantine proposer

**Comment by milosevic** [comment]
I don't see algorithmic problem. Upon committing some blockID, in case we don't have the corresponding block, we set up to get it: 
https://github.com/tendermint/tendermint/blob/develop/consensus/state.go#L1136-L1141

After line 1141 is executed, we shouldn't let wrong block part to be added to cs.ProposalBlockParts. 
See: https://github.com/tendermint/tendermint/blob/68a0b3f95b331e04efabff9817fd60b8c1ce3b28/types/part_set.go#L179-L205

So I don't understand how we can end up in state where proposal-block=3C0555F961CEE5E150B145A5EADA10E6BD41329D after we have set up commit-block=9DC0B9395F81CB0367517E57F7B6205535091B21. 

Can we have two blockID with the same PartSetHeader but different Hash? I am a bit confused with the fact that BlockID contains both block hash and PartSetHeader and that the Proposal only contains PartSetHeader. Shouldn't PartSetHeader be enough to uniquely identify ProposalBlock?

**Comment by ebuchman** [comment]
Here are the actual blocks that got committed at height 162664 and 162665: https://gist.github.com/ebuchman/951ceb3ee8c7566ddfda179189970a78

You can see block 162664 is different than the one proposed here (it includes a different LastCommit).

Also you can see from block 162665 that block 162664 was committed in round 3.

It should not be possible to have the same PartSetHeader but different Hash. PartSetHeader is just another way of hashing the block (ie. merkle tree after serializing the entire block, rather than just taking a merkle tree of the header). Though I agree, it's confusing.

You seem to be correct that it shouldn't be possible to get in the state chris got in so obviously we're missing something ...

**Comment by ebuchman** [comment]
Aha!

I believe the culprit is in fact https://github.com/tendermint/tendermint/blob/v0.19.1/consensus/state.go#L1298

Specifically, if we received the block part from ourselves, then we'll have `verify=false`, which was a performance optimization to avoid merkle hashing a block we created.

But in this case, since the block we're supposed to commit is different than our proposal, the lack of verification messed everything up.

So really we should remove that flag and just always verify and this problem should go away.

Shazam!

**Comment by ebuchman** [comment]
Fixed on develop and in Tendermint v0.19.4 (currently in rc0)

Should be fixed in gaia here: https://github.com/cosmos/cosmos-sdk/pull/1019

**Comment by ebuchman** [comment]
@cwgoes looks like we have seen this before in #861 ! ^^
